sklearn/_engine/base.py

```diff
@@ -112,3 +112,37 @@ def get_engine_classes(engine_name, default, verbose=False):
                 f"trying engine {engine_class.__module__}.{engine_class.__qualname__} ."
             )
         yield provider, engine_class
+
+
+def convert_attributes(method):
+    """Convert estimator attributes after calling the decorated method.
+
+    The attributes of an estimator can be stored in "engine native" types
+    (default) or "scikit-learn native" types. This decorator will call the
+    engine's conversion function when needed. Use this decorator on methods
+    that set estimator attributes.
+    """
+
+    @wraps(method)
+    def wrapper(self, *args, **kwargs):
+        r = method(self, *args, **kwargs)
+        convert_attributes = get_config()["engine_attributes"]
+
+        if convert_attributes == "sklearn_types":
+            engine = self._engine_class
+            for name, attribute in vars(self).items():
+                # All attributes are passed to the engine, which can
+                # either convert the value (engine specific types) or
+                # return it as is (native Python types)
+                converted = engine.convert_to_sklearn_types(name, attribute)
```

Comment:
```suggestion
            for name, value in vars(self).items():
                # All attributes are passed to the engine, which can
                # either convert the value (engine specific types) or
                # return it as is (native Python types)
                converted = engine.convert_to_sklearn_types(name, value)
```

Reply:
We could additionally allow the engine to implement an optional `engine.convert_estimator_to_sklearn_types` method if we want to give it the flexibility to do the attribute looping and `setattr` logic by itself.

But this probably qualifies as a YAGNI for now. Let's just keep this possibility in mind for later if find the default order returned by `vars(self).items()` annoying for some estimators.

Reply:
Another potential limitation of this design would be the handling of the case when the `EngineAwareMixin` class is used on a meta-estimator. But let's keep this for later.

Reply:
> Another potential limitation of this design would be the handling of the case when the `EngineAwareMixin` class is used on a meta-estimator. But let's keep this for later.

Just to double check: you are thinking of the estimator(s) that are stored as attributes of the meta-estimator right?